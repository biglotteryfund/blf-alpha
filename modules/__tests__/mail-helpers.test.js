/* eslint-env jest */
'use strict';
const nodemailer = require('nodemailer');
const { getSendAddress, buildMailOptions } = require('../mail-helpers');

describe('getSendAddress', () => {
    const expectedDefault = `noreply@biglotteryfund.org.uk`;
    const expectedInternal = `noreply@blf.digital`;

    it('should return default send from address for external send to addresses', () => {
        expect(getSendAddress('example@example.com')).toBe(expectedDefault);
        expect(getSendAddress('example@gmail.com')).toBe(expectedDefault);
    });

    it('should return internal send from address for internal send to addresses', () => {
        expect(getSendAddress('example@biglotteryfund.org.uk')).toBe(expectedInternal);
        // Assert against similar looking but incorrect emails to test for false positives
        expect(getSendAddress('example@biggerlotteryfund.org.uk')).toBe(expectedDefault);
        expect(getSendAddress('example@biggestlotteryfund.org.uk')).toBe(expectedDefault);
        expect(getSendAddress('biglotteryfund.org.uk@example.com')).toBe(expectedDefault);
    });
});

async function mockEmail(mailOptions) {
    const transport = nodemailer.createTransport({
        streamTransport: true,
        newline: 'unix',
        buffer: true
    });

    const info = await transport.sendMail(mailOptions);
    // Remove generated properties from message
    info.message = info.message
        .toString()
        .replace(/Message-ID: .*\n/, '')
        .replace(/Date: .*\n/, '')
        .replace(/ boundary=.*\n/, '')
        .replace(/----.*\n/gm, '');
    return info;
}

describe('buildMailOptions', () => {
    it('should build mail options for a text email', async () => {
        const mailOptions = buildMailOptions({
            subject: 'Test',
            sendTo: { address: 'example@example.com' },
            content: 'This is a test'
        });

        expect(mailOptions).toMatchSnapshot();
        const info = await mockEmail(mailOptions);
        expect(info.envelope).toMatchSnapshot();
        expect(info.message).toMatchSnapshot();
    });

    it('should build mail options for a text email with a name', async () => {
        const mailOptions = buildMailOptions({
            subject: 'Test',
            sendTo: { name: 'Example Person', address: 'example@example.com' },
            content: 'This is a test'
        });

        expect(mailOptions).toMatchSnapshot();
        const info = await mockEmail(mailOptions);
        expect(info.envelope).toMatchSnapshot();
        expect(info.message).toMatchSnapshot();
    });

    it('should build mail options for a html email', async () => {
        const mailOptions = buildMailOptions({
            subject: 'Test',
            sendTo: { address: 'example@example.com' },
            type: 'html',
            content: '<p>This is a test</p>'
        });

        expect(mailOptions).toMatchSnapshot();
        const info = await mockEmail(mailOptions);
        expect(info.envelope).toMatchSnapshot();
        expect(info.message).toMatchSnapshot();
    });

    it('should build mail options for an internal html email', async () => {
        const mailOptions = buildMailOptions({
            subject: 'Test',
            sendTo: { address: 'example@biglotteryfund.org.uk' },
            type: 'html',
            content: '<p>This is a test</p>'
        });

        expect(mailOptions).toMatchSnapshot();
        const info = await mockEmail(mailOptions);
        expect(info.envelope).toMatchSnapshot();
        expect(info.message).toMatchSnapshot();
    });
});
