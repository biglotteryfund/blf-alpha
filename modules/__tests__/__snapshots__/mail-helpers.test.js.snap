// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildMailOptions should build mail options for a html email 1`] = `
Object {
  "from": "Big Lottery Fund <noreply@biglotteryfund.org.uk>",
  "html": "<p>This is a test</p>",
  "subject": "Test",
  "text": "This is a test",
  "to": Object {
    "address": "example@example.com",
  },
}
`;

exports[`buildMailOptions should build mail options for a html email 2`] = `
Object {
  "from": "noreply@biglotteryfund.org.uk",
  "to": Array [
    "example@example.com",
  ],
}
`;

exports[`buildMailOptions should build mail options for a html email 3`] = `
"Content-Type: multipart/alternative;
From: Big Lottery Fund <noreply@biglotteryfund.org.uk>
To: example@example.com
Subject: Test
MIME-Version: 1.0

Content-Type: text/plain
Content-Transfer-Encoding: 7bit

This is a test
Content-Type: text/html
Content-Transfer-Encoding: 7bit

<p>This is a test</p>
"
`;

exports[`buildMailOptions should build mail options for a text email 1`] = `
Object {
  "from": "Big Lottery Fund <noreply@biglotteryfund.org.uk>",
  "subject": "Test",
  "text": "This is a test",
  "to": Object {
    "address": "example@example.com",
  },
}
`;

exports[`buildMailOptions should build mail options for a text email 2`] = `
Object {
  "from": "noreply@biglotteryfund.org.uk",
  "to": Array [
    "example@example.com",
  ],
}
`;

exports[`buildMailOptions should build mail options for a text email 3`] = `
"Content-Type: text/plain
From: Big Lottery Fund <noreply@biglotteryfund.org.uk>
To: example@example.com
Subject: Test
Content-Transfer-Encoding: 7bit
MIME-Version: 1.0

This is a test
"
`;

exports[`buildMailOptions should build mail options for a text email with a name 1`] = `
Object {
  "from": "Big Lottery Fund <noreply@biglotteryfund.org.uk>",
  "subject": "Test",
  "text": "This is a test",
  "to": Object {
    "address": "example@example.com",
    "name": "Example Person",
  },
}
`;

exports[`buildMailOptions should build mail options for a text email with a name 2`] = `
Object {
  "from": "noreply@biglotteryfund.org.uk",
  "to": Array [
    "example@example.com",
  ],
}
`;

exports[`buildMailOptions should build mail options for a text email with a name 3`] = `
"Content-Type: text/plain
From: Big Lottery Fund <noreply@biglotteryfund.org.uk>
To: Example Person <example@example.com>
Subject: Test
Content-Transfer-Encoding: 7bit
MIME-Version: 1.0

This is a test
"
`;

exports[`buildMailOptions should build mail options for an internal html email 1`] = `
Object {
  "from": "Big Lottery Fund <noreply@blf.digital>",
  "html": "<p>This is a test</p>",
  "subject": "Test",
  "text": "This is a test",
  "to": Object {
    "address": "example@biglotteryfund.org.uk",
  },
}
`;

exports[`buildMailOptions should build mail options for an internal html email 2`] = `
Object {
  "from": "noreply@blf.digital",
  "to": Array [
    "example@biglotteryfund.org.uk",
  ],
}
`;

exports[`buildMailOptions should build mail options for an internal html email 3`] = `
"Content-Type: multipart/alternative;
From: Big Lottery Fund <noreply@blf.digital>
To: example@biglotteryfund.org.uk
Subject: Test
MIME-Version: 1.0

Content-Type: text/plain
Content-Transfer-Encoding: 7bit

This is a test
Content-Type: text/html
Content-Transfer-Encoding: 7bit

<p>This is a test</p>
"
`;
